### PR TITLE
Add blog post about k8s runtime observability

### PR DESCRIPTION
**Pull request by danielbdias**
Hi folks! 

Last week I was doing a PoC on instrumenting Kubernetes with OpenTelemetry traces with [Tracetest](https://tracetest.io/) team, and we thought that it could be good to blog about it and wanted to share it with you in this blog post.

What are your thoughts about it? Do you think it makes sense?

---

Preview:

- ~https://deploy-preview-2779--opentelemetry.netlify.app/blog/2023/creating-k8s-cluster-with-runtime-observability/~
- https://deploy-preview-2779--opentelemetry.netlify.app/blog/2023/k8s-runtime-observability/

### CODE REVIEW

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+that update internal resources to add more functionalities to Kubernetes.
+
+[Tracetest](https://tracetest.io/) is interested in seeing how
+applying trace-based testing techniques to traces from the core Kubernetes
+engine can be leveraged to verify expected functionality. With the current level
+of observability being exposed by Kubernetes, we feel it is a bit early to work
+to enable the execution of `kubectl` commands as one of the supported triggers
+for a trace-based test. We did, however, want to publish our findings in order
+to help others interested in seeing the current state of distributed tracing in
+the Kubernetes engine.
```
Comment (svrnm) [inline content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md:462]:
```suggestion
```
I don't think this section is relevant for the OpenTelemetry blog. Can we remove it?

**Review by svrnm — CHANGES_REQUESTED** [review]
Thank you @danielbdias! I like what you wrote down here. I provided some feedback, please review.

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 2]
```diff
@@ -0,0 +1,474 @@
+---
+title:
+  'Creating a Kubernetes Cluster with Runtime Observability Using k3d and
+  OpenTelemetry'
+linkTitle: 'Creating a Kubernetes Cluster with Runtime Observability'
```
[the remark this answers is not in the document]
Comment (chalin) [reply content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md:5]: Nit: FYI, you don't need any quotes around those strings

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+### Creating a Kubernetes Cluster with Runtime Observability
+
+With the observability environment set up, create the configuration files to
+enable OpenTelemetry tracing in `kube-apiserver`, `kubelet`, and `containerd`.
+
+Inside the `kubetracing` folder, create a subfolder called `config` that will
+have the following two files.
+
+First, the
+`[apiserver-tracing.yaml](https://github.com/kubeshop/tracetest/blob/main/examples/tracetesting-kubernetes/kubetracing/config/apiserver-tracing.yaml)`,
```
Comment (chalin) [inline content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md:174]:
```suggestion
[apiserver-tracing.yaml](https://github.com/kubeshop/tracetest/blob/main/examples/tracetesting-kubernetes/kubetracing/config/apiserver-tracing.yaml),
```

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+kind: KubeletConfiguration
+featureGates:
+  KubeletTracing: true
+tracing:
+  endpoint: host.k3d.internal:4317
+  samplingRatePerMillion: 1000000
+```
+
+Returning to the `kubetracing` folder, create the last file,
+`[config.toml.tmpl](https://github.com/kubeshop/tracetest/blob/main/examples/tracetesting-kubernetes/kubetracing/config.toml.tmpl)`,
```
Comment (chalin) [inline content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md:207]:
```suggestion
[config.toml.tmpl](https://github.com/kubeshop/tracetest/blob/main/examples/tracetesting-kubernetes/kubetracing/config.toml.tmpl),
```

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+which is a template file used by `k3s` to configure `containerd`. This file is
+similar to the default configuration that `k3s` uses, with two more sections at
+the end of the file that configures `containerd` to send traces.
+
+```toml
+# File generated by {{ .Program }}. DO NOT EDIT. Use config.toml.tmpl instead.
```
Comment (chalin) [inline content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md:213]: Is this comment line really necessary?

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+  samplingRatePerMillion: 1000000
+```
+
+Returning to the `kubetracing` folder, create the last file,
+`[config.toml.tmpl](https://github.com/kubeshop/tracetest/blob/main/examples/tracetesting-kubernetes/kubetracing/config.toml.tmpl)`,
+which is a template file used by `k3s` to configure `containerd`. This file is
+similar to the default configuration that `k3s` uses, with two more sections at
+the end of the file that configures `containerd` to send traces.
+
+```toml
```
Comment (chalin) [inline content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md:212]: I think that using `liquid` might make for a better result?

```suggestion
```liquid
```

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+Kubernetes control plane is running at https://0.0.0.0:60503
+CoreDNS is running at https://0.0.0.0:60503/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy
+Metrics-server is running at https://0.0.0.0:60503/api/v1/namespaces/kube-system/services/https:metrics-server:https/proxy
+```
+
+Going back to the logs of the observability environment, you should see some
+spans of internal Kubernetes operations being published in OpenTelemetry
+Collector, like this:
+
+```bash
```
Comment (chalin) [inline content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md:372]: This isn't `bash` code. Consider dropping the language specifier?

```suggestion
```
```

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+```
+
+### Testing the Cluster Runtime
+
+With the Observability environment and the Kubernetes cluster set up, you can
+now trigger commands against Kubernetes and see traces of these actions in
+Jaeger.
+
+Open the browser, and navigate to the Jaeger UI located at
+[http://localhost:16686/search](http://localhost:16686/search). You’ll see that
```
Comment (chalin) [inline content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md:407]:
```suggestion
<http://localhost:16686/search>. You’ll see that
```

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+Choose `apiserver` and click on **"Find Traces”.** Here you see traces from the
+Kubernetes control plane:
+
+![Jaeger screen showing a list of spans found for apiserver](https://res.cloudinary.com/djwdcmwdz/image/upload/v1684855720/Blogposts/creating-kubernetes-cluster-with-observability/screely-1684855412765_xe7rnr.png)
+
+Let’s run a sample command against Kubernetes with `kubectl`, like running an
+echo:
+
+```bash
+kubectl run -it --rm --image=alpine echo-command -- echo Hi
```
Comment (chalin) [inline content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md:421]:
```suggestion
```console
$ kubectl run -it --rm --image=alpine echo-command -- echo Hi
```

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+# warning: couldn't attach to pod/echo-command, falling back to streaming logs: unable to upgrade connection: container echo-command not found in pod echo-command_default
+# Hi
+# pod "echo-command" deleted
+```
+
+And now, open Jaeger again, choose the `kubelet` service, operation `syncPod`,
+and add the tag `k8s.pod=default/echo-command`, you should be able to see spans
+related to this pod:
+
+![Jaeger screen showing a list of spans found for the syncPod operation on kubelet service](https://res.cloudinary.com/djwdcmwdz/image/upload/v1684855720/Blogposts/creating-kubernetes-cluster-with-observability/screely-1684855456007_flekwn.png)
```
Comment (chalin) [inline content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md:434]: Images should be local. For help on how to set that up see: https://github.com/open-telemetry/opentelemetry.io/blob/main/README.md#fork--write-locally

**Review by chalin — CHANGES_REQUESTED** [review]
Like for other blog posts, images should be local. See inline comments for details.

I'll be making another pass soon.

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+With contributions from [Sebastian Choren](https://github.com/schoren),
+[Adnan Rahić](https://github.com/adnanrahic) and
+[Ken Hamric](https://github.com/kdhamric).
+
+## Creating a Kubernetes Cluster with Runtime Observability Using k3d and OpenTelemetry
+
+[Kubernetes](https://kubernetes.io/) is an open-source system widely used in the
+cloud-native landscape to provide ways to deploy and scale containerized
+applications in the cloud. Its ability to observe logs and metrics is well-known
+and documented, but its observability regarding application traces is new.
```
[the remark this answers is not in the document]
Comment (danielbdias) [reply content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md]: It makes sense! I've added a reference to the official Kubernetes docs here: https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-usage-monitoring/

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+  [beta in v1.27](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.27.md)
+  (Apr. 2023).
+- Kubelet tracing was released as
+  [alpha in v1.25](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/2831-kubelet-tracing/kep.yaml#L29)
+  (Aug. 2022) and
+  [beta in v1.27](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.27.md)
+  (Apr. 2023).
+
+In investigating the current state of tracing with Kubernetes, we found very few
+articles documenting how to enable it. We decided to document our findings and
```
[the remark this answers is not in the document]
Comment (danielbdias) [reply content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md]: I added a reference to a Kubernetes blog post (which inspired me to write this one) that gives an idea of this implementation. I'm referencing it also at the References section at the end of the article.

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+    sampling_percentage: 100
+  batch:
+    timeout: 100ms
+exporters:
+  logging:
+    logLevel: debug
+  jaeger:
+    endpoint: jaeger:14250
+    tls:
+      insecure: true
```
[the remark this answers is not in the document]
Comment (danielbdias) [reply content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md]: Makes sense! I've updated the exporter name and port to make it explicit, and also the `docker-compose.yaml` configuration.

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md (hunk 1 above)
Comment (danielbdias) [reply content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md]: Sure, we can remove it! But, I would love to praise my colleagues who helped me research and implement this use case. How about replacing it with something like this:

```
As a team focused on building an open source tool in the observability space,
the opportunity to help the overall OpenTelemetry community was important to us.
That’s why we were researching finding new ways of collecting traces from the
core Kubernetes engine. With the current level of observability being exposed by
Kubernetes we wanted to publish our findings in order to help others interested
in seeing the current state of distributed tracing in the Kubernetes engine.
Daniel Dias and Sebastian Choren are working on Tracetest, an open-source tool
that allows you to develop and test your distributed system with OpenTelemetry.
It works with any OTel compatible system and enables trace–based tests to be
created. Check it out at https://github.com/kubeshop/tracetest.
```

Do you think it makes sense? If not, the is no problem, I can remove this paragraph.

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md [hunk 14]
```diff
@@ -0,0 +1,472 @@
...
+author: '[Daniel Dias](https://github.com/danielbdias) (Tracetest)'
+# canonical_url: http://somewhere.else/ # TODO: if this blog post has been posted somewhere else already, uncomment & provide the canonical URL here.
+body_class: otel-with-contributions-from
+---
+
+With contributions from [Sebastian Choren](https://github.com/schoren),
+[Adnan Rahić](https://github.com/adnanrahic) and
+[Ken Hamric](https://github.com/kdhamric).
+
+## Creating a Kubernetes Cluster with Runtime Observability Using k3d and OpenTelemetry
```
[the remark this answers is not in the document]
Comment (danielbdias) [reply content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md]: Ok! Thanks!

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md (hunk 7 above)
Comment (danielbdias) [reply content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md]: It makes sense, I've dropped the language specifier.

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md (hunk 5 above)
Comment (danielbdias) [reply content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md]: No, I'm removing it.

**Comment by danielbdias** [comment]
@svrnm and @chalin, thanks for your feedback! I've pushed a new version considering your comments.

File: content/en/blog/2023/k8s-runtime-observability/index.md [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+
+With the Observability environment and the Kubernetes cluster set up, you can
+now trigger commands against Kubernetes and see traces of these actions in
+Jaeger.
+
+Open the browser, and navigate to the Jaeger UI located at
+<http://localhost:16686/search>. You’ll see that the `apiserver`, `containerd`,
+and `kubelet` services are publishing traces:
+
+![Jaeger screen with services dropdown open showing apiserver, containerd and kubelet services as options](./k8s-services-reported-on-jaeger.png)
```
Comment (chalin) [inline content/en/blog/2023/k8s-runtime-observability/index.md:409]:
```suggestion
![Jaeger screen with services dropdown open showing apiserver, containerd and kubelet services as options](k8s-services-reported-on-jaeger.png)
```

File: content/en/blog/2023/k8s-runtime-observability/index.md [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+
+And now, open Jaeger again, choose the `kubelet` service, operation `syncPod`,
+and add the tag `k8s.pod=default/echo-command`, you should be able to see spans
+related to this pod:
+
+![Jaeger screen showing a list of spans found for the syncPod operation on kubelet service](./syncpod-operations-on-kubelet.png)
+
+Expanding one trace, you’ll see the operations that created this pod:
+
+![Jaeger screen showing a single syncPod expanded](./single-syncpod-expanded.png)
```
Comment (chalin) [inline content/en/blog/2023/k8s-runtime-observability/index.md:437]:
```suggestion
![Jaeger screen showing a single syncPod expanded](single-syncpod-expanded.png)
```

File: content/en/blog/2023/k8s-runtime-observability/index.md [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+in seeing the current state of distributed tracing in the Kubernetes engine.
+Daniel Dias and Sebastian Choren are working on Tracetest, an open-source tool
+that allows you to develop and test your distributed system with OpenTelemetry.
+It works with any OTel compatible system and enables trace–based tests to be
+created. Check it out at https://github.com/kubeshop/tracetest.
+
+You can find the examples that were used in this article
+[here](https://github.com/kubeshop/tracetest/tree/main/examples/tracetesting-kubernetes/kubetracing)
+and a brief description of them
+[here](https://github.com/kubeshop/tracetest/blob/main/examples/tracetesting-kubernetes/setup-k8s-with-k3d.md).
```
Comment (chalin) [inline content/en/blog/2023/k8s-runtime-observability/index.md:466]: Avoid "here" as link text:

```suggestion
The
[example sources](https://github.com/kubeshop/tracetest/tree/main/examples/tracetesting-kubernetes/kubetracing)
used in this article, and
[setup instructions](https://github.com/kubeshop/tracetest/blob/main/examples/tracetesting-kubernetes/setup-k8s-with-k3d.md)
are available from the Tracetest repo.
```

File: content/en/blog/2023/k8s-runtime-observability/index.md [hunk 18]
```diff
@@ -0,0 +1,474 @@
...
+- [k3d](https://k3d.io/): a wrapper to run [k3s](https://k3s.io/) (a lightweight
+  Kubernetes distribution) with Docker
+- [kubectl](https://kubernetes.io/docs/reference/kubectl/): a Kubernetes CLI to
+  interact with clusters
+
+## Setting up an Observability Stack to Monitor Traces
+
+To set up the observability stack, you’ll run the
+[OpenTelemetry Collector](../../../docs/collector/) (OTel
+Collector), a tool that receives telemetry data from different apps and sends it
```
Comment (chalin) [inline content/en/blog/2023/k8s-runtime-observability/index.md:77]:
```suggestion
To set up the observability stack, you’ll run the OpenTelemetry (OTel)
[Collector](/docs/collector/), a tool that receives telemetry data from different apps and sends it
```

File: content/en/blog/2023/k8s-runtime-observability/index.md [hunk 19]
```diff
@@ -0,0 +1,474 @@
+---
+title: Creating a Kubernetes Cluster with Runtime Observability
+linkTitle: K8s Runtime Observability
+date: 2023-05-23
+author: '[Daniel Dias](https://github.com/danielbdias) (Tracetest)'
+body_class: otel-with-contributions-from
```
Comment (chalin) [inline content/en/blog/2023/k8s-runtime-observability/index.md:6]:
```suggestion
body_class: otel-with-contributions-from
spelling: cSpell:ignore apiserver choren tracetest kube kubelet containerd
```

**Review by chalin — COMMENTED** [review]
See inline comments.

File: content/en/blog/2023/k8s-runtime-observability/index.md (hunk 17 above)
Comment (danielbdias) [reply content/en/blog/2023/k8s-runtime-observability/index.md]: It makes sense, thanks!

File: content/en/blog/2023/k8s-runtime-observability/index.md [hunk 20]
```diff
@@ -0,0 +1,475 @@
...
+Choose `apiserver` and click on **"Find Traces”.** Here you see traces from the
+Kubernetes control plane:
+
+![Jaeger screen showing a list of spans found for apiserver](spans-found-for-apiserver.png)
+
+Let’s run a sample command against Kubernetes with `kubectl`, like running an
+echo:
+
+```bash
+kubectl run -it --rm --restart=Never --image=alpine echo-command -- echo hi
```
Comment (chalin) [inline content/en/blog/2023/k8s-runtime-observability/index.md:420]:
```suggestion
```console
$ kubectl run -it --rm --restart=Never --image=alpine echo-command -- echo hi
```

**Review by chalin — APPROVED** [review]
LGTM (I'd love to make a full pass over the prose, but won't have time -- I'll defer to @svrnm and @cartermp)

**Comment by chalin** [comment]
Of course, you need to address the check failures.

**Comment by danielbdias** [comment]
Hi folks, thanks for your feedback! 

I've addressed the pipeline failures. It should be working now. @chalin, @svrnm, and @cartermp, feel free to add any comments about the changes and existing text.

Thanks!

File: content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (svrnm) [reply content/en/blog/2023/creating-k8s-cluster-with-runtime-observability.md]: thank you

**Review by cartermp — APPROVED** [review]
This is great! I think we should merge this and have it set to publish for next week.

@danielbdias could set the `date` at the top to next Monday? Thanks!

**Comment by danielbdias** [comment]
@cartermp done! Thanks!

**Pull request merged**